File: v4/rust/iprange-livedb/src/recovery/structured_build.rs
//! Canonical immutable output from one structured recovery analysis.

use crate::cancellation::CancellationToken;
use crate::contract::{MetaV4, StructureKind, ValueKind};
use crate::error::{Error, Result};
use crate::immutable_output::Builder;
use crate::mapping::Mapping;

use super::construction::{Construction, Failure};
use super::indirect_build::{self, Mode, OutputContext};
use super::membership_output::MembershipKey;
use super::range_build::{build_ranges, BuildResult};
use super::range_components::Components;
use super::report::RecoverySink;
use super::structure_table::StructureIndex;
use super::structured_output::{NetworkEnrichmentV1Output, StructuredKey};
use super::RecoveryBudget;

struct Structured;

impl Mode for Structured {
    const VALUE_KIND: ValueKind = ValueKind::Structured;

    fn check_structures(structures: Option<&StructureIndex>) -> Result<()> {
        let structures =
            structures.ok_or(Error::Corrupt("structured recovery has no structure index"))?;
        if structures.kind() != StructureKind::NetworkEnrichmentV1 {
            return Err(Error::UnsupportedStructure(structures.kind() as u8));
        }
        Ok(())
    }

    fn output<K, S>(context: OutputContext<'_, '_, '_, S>) -> BuildResult
    where
        K: MembershipKey + StructuredKey,
        S: RecoverySink,
    {
        let structures = context
            .structures
            .expect("structured recovery index was checked");
        let policy = NetworkEnrichmentV1Output::<S, K>::new(
            context.request.mapping,
            context.request.meta,
            context.memberships,
            structures,
            context.tables,
            context.builder,
            context.reporter,
        );
        let mut output = Components::new(context.request.cancellation, policy);
        build_ranges(context.request, context.pages, &mut output)
    }
}

#[allow(clippy::result_large_err)]
pub(crate) fn construct<S: RecoverySink>(
    mapping: &Mapping,
    source_meta: MetaV4,
    builder: Builder,
    budget: &RecoveryBudget,
    cancellation: &CancellationToken,
    sink: &mut S,
) -> std::result::Result<Construction, Failure> {
    indirect_build::construct::<Structured, S>(
        mapping,
        source_meta,
        builder,
        budget,
        cancellation,
        sink,
    )
}

#[cfg(test)]
#[path = "structured_tests.rs"]
mod tests;

File: v4/go/internal/recovery/structured_build.go
package recovery

// Canonical immutable output from one structured recovery analysis
// (Rust recovery/structured_build.rs): the structured mode proves the
// network-enrichment structure index and streams the structured-range
// policy over the shared source-ID lookups.

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/mapping"
	"github.com/firehol/iprange/v4/go/internal/writer"
)

// structuredConstruct builds one canonical structured output from one
// recovery source (Rust structured_build::construct).
func structuredConstruct(m *mapping.Mapping, sourceMeta format.Meta, builder *writer.OutputBuilder, budget *RecoveryBudget, check func() error, sink RecoverySink) (*Construction, *constructionFailure) {
	return indirectConstruct(indirectMode{
		kind: format.ValueKindStructured,
		checkStructures: func(structures *structureIndex) error {
			if structures == nil {
				return corruptError("structured recovery has no structure index")
			}
			if structures.kind != format.StructureKindNetworkEnrichmentV1 {
				return &format.Error{Code: format.CodeUnsupportedStructure, Detail: "recovery structure kind is unsupported"}
			}
			return nil
		},
		output: func(context indirectOutputContext) (*scratchCleanup, *rangeBuildFailure) {
			codec, ok := indirectCodec(context.request.meta.AddressFamily)
			if !ok {
				return nil, &rangeBuildFailure{cause: &format.Error{Code: format.CodeFormatInvalid, Detail: "recovery structured output family is invalid"}}
			}
			structures := context.structures
			if structures == nil {
				return nil, &rangeBuildFailure{cause: corruptError("structured recovery has no structure index")}
			}
			policy := &structuredOutput{
				mapping:     context.request.mapping,
				meta:        context.request.meta,
				memberships: context.memberships,
				structures:  structures,
				tables:      context.tables,
				builder:     context.builder,
				rep:         context.rep,
				family:      context.request.meta.AddressFamily,
			}
			return buildRanges(codec, context.request, context.pages, &components{check: context.request.check, codec: codec, policy: policy})
		},
	}, m, sourceMeta, builder, budget, check, sink)
}
